pkg/handlers/handler_base.go

```diff
@@ -0,0 +1,171 @@
+	return HandlerBase{
+		logger: logger,
+		clock:  clock.New(),
+	}
+}
+
+// HandlerBase is for shared handler utilities
+type HandlerBase struct {
+	logger *zap.Logger
+	clock  clock.Clock
+}
+
+type errorItem struct {
+	Field   string `json:"field"`
+	Message string `json:"message"`
+}
+
+// errorResponse contains the structure of error for a http response
+type errorResponse struct {
+	Errors  []errorItem `json:"errors"`
+	Code    int         `json:"code"`
+	Message string      `json:"message"`
+	TraceID uuid.UUID   `json:"traceID"`
+}
+
+func newErrorResponse(code int, message string, traceID uuid.UUID) errorResponse {
+	return errorResponse{
+		Errors:  []errorItem{},
+		Code:    code,
+		Message: message,
+		TraceID: traceID,
+	}
+}
+
+func (r *errorResponse) withMap(errMap map[string]string) {
+	for k, v := range errMap {
+		r.Errors = append(r.Errors, errorItem{
+			Field:   k,
+			Message: v,
+		})
+	}
+}
+
+// WriteErrorResponse writes a response for a given application error
+func (b HandlerBase) WriteErrorResponse(ctx context.Context, w http.ResponseWriter, appErr error) {
+	logger, ok := appcontext.Logger(ctx)
+	if !ok {
+		logger = b.logger
+	}
+
+	traceID, ok := appcontext.Trace(ctx)
+	if !ok {
+		traceID = uuid.New()
+		logger.With(zap.String("traceID", traceID.String()))
+	}
+
+	// get code and response
+	var code int
+	var response errorResponse
+	switch appErr := appErr.(type) {
+	case *apperrors.UnauthorizedError:
+		// 4XX errors are not logged as errors, but are for client
+		logger.Info("Returning unauthorized response from handler", zap.Error(appErr))
+		code = http.StatusUnauthorized
+		response = newErrorResponse(
+			http.StatusUnauthorized,
+			"Unauthorized",
+			traceID,
+		)
+	case *apperrors.QueryError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Something went wrong",
+			traceID,
+		)
+	case *apperrors.NotificationError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Failed to send notification",
+			traceID,
+		)
+	case *apperrors.ExternalAPIError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Something went wrong",
+			traceID,
+		)
+	case *apperrors.ContextError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Something went wrong",
+			traceID,
+		)
+	case *apperrors.ValidationError:
+		logger.Info("Returning bad request error from handler", zap.Error(appErr))
+		code = http.StatusBadRequest
```

Comment:
These should also be 422 Unprocessable Entity because the action can't be taken with the data as is.

Reply:
yeah, I think we need to add 400 elsewhere for malformed json payloads or invalid types (like uuid vs string)